### ISSUE TITLE
[dynamo] Data-dep branching on torch.all(mask==1) mamba/linear-attn fast-path (20 models, 204 breaks)

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#78]
**Repro status:** Reproduces on torch 2.12.0.dev20260407+cu128 (nightly, verified 2026-05-09 ~17:44 UTC).

## Original failure report

20 mamba/linear-attention family models (Jamba*, FalconH1*, etc.) produce 204 graph breaks from a "mask all-ones fast-path" pattern: `if attention_mask is not None and torch.all(attention_mask == 1): use_fast_path()`. The `torch.all(mask == 1)` branch is data-dependent — Dynamo can't trace.

**Test command (sweep):** `python tools/run_experiment.py sweep ...` (cohort: 20 mamba/linear-attention models, animesh-fullgraph 2026-04-28).

**Symptom captured:** `Data-dependent branching` (gb0170). Cited stack frame for the baseline model: `transformers/models/jamba/modeling_jamba.py:755`.

### Affected Models

20 models in `sweep_results/experiments/animesh-fullgraph-2026-04-28/explain_results.json` (Jamba*, FalconH1*, etc. — full list in the sweep evidence).

<details><summary>Verification signal (original)</summary>

`{"kind": "stderr_contains", "fragment": "Data-dependent branching"}`
</details>

## Pattern

Cited call site at `transformers/models/jamba/modeling_jamba.py:754-755` (mask all-ones fast-path):

```text
if attention_mask is not None and torch.all(attention_mask == 1):
    mamba_mask = None  # fast-path
else:
    mamba_mask = attention_mask
```

`torch.all(attention_mask == 1)` is a data-dependent boolean — Dynamo refuses to branch on it.

## Minimal reproducer (MRE)

15 lines, **no transformers dependency** — the gap is generic; any `torch.all(tensor == value)` inside compiled forward triggers it.

```python repro=true
import os
os.environ.setdefault("TORCH_LOGS", "graph_breaks")
import torch
import torch.nn as nn

class JambaMaskFastPath(nn.Module):
    def forward(self, attention_mask):
        if attention_mask is not None and torch.all(attention_mask == 1):
            mamba_mask = None
        else:
            mamba_mask = attention_mask
        return mamba_mask

m = JambaMaskFastPath()
compiled = torch.compile(m, backend="eager", dynamic=True)
attention_mask = torch.ones(1, 4, dtype=torch.long)
out = compiled(attention_mask)
print("ran, out =", out)
```

<details><summary>Verification signal (MRE)</summary>

`{"kind": "stderr_contains", "fragment": "Data-dependent branching"}`
</details>

## Source

- Sweep results: `sweep_results/experiments/animesh-fullgraph-2026-04-28/explain_results.json` (20 mamba/linear-attn rows)
- **Provenance anchor:** `transformers/models/jamba/modeling_jamba.py:755`
- Built off PR pytorch/pytorch#181552
- Transformers: 5.6.2

---
<sub>Body rewritten 2026-05-09 by file-issue + mre subagent dogfood. MRE deduced from cited fast-path branch, verified on torch 2.12 nightly. Title verified accurate.</sub>

**Comment by penguinwu** [comment]
After today's corpus modellibs upgrade to `transformers 5.8.0` (this repo, commit `e0301da`), two additional model families join this issue's territory: **GraniteMoeHybrid** (4 pairs: `Model` + `ForCausalLM` × `eval` + `train`) and **MiniCPMV4_6** (4 pairs: `Model` + `ForConditionalGeneration` × `eval` + `train`). Both now break on the `torch.all(attention_mask == 1)` mask fast-path pattern under transformers 5.8.0:

- **GraniteMoeHybrid:** `transformers/models/granitemoehybrid/modeling_granitemoehybrid.py:1219` — `attention_mask is not None and torch.all(attention_mask == 1)` in `_update_mamba_mask`. 8 graph breaks per pair (4 pairs total: 2 modes × 2 model classes).
- **MiniCPMV4_6:** `transformers/models/qwen3_5/modeling_qwen3_5.py:1312` (the linear-attn variant of the same pattern). 3-4 graph breaks per pair.
- **Already tracked:** `transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py:1437` (per the umbrella #122 audit).

Total scope expansion on tx 5.8.0: **+8 model×mode pairs, ~48 graph breaks**. Note: on transformers 5.6.2 the same GraniteMoeHybrid pairs broke on a different pattern (`_local_scalar_dense` from `.tolist()` at modeling_granitemoehybrid.py:924, #55 territory) — the issue this comment posts on (#78) is the corrected attribution under the upstream tx 5.8.0 source.
